client/client_test.go

```diff
@@ -7902,6 +7903,75 @@ func testMultipleCacheExports(t *testing.T, sb integration.Sandbox) {
 	ensureFileContents(t, filepath.Join(destDir, "unique"), string(uniqueFile))
 }
 
+// https://github.com/moby/buildkit/issues/3148
+func testMountStubsTimestamp(t *testing.T, sb integration.Sandbox) {
+	c, err := New(sb.Context(), sb.Address())
+	require.NoError(t, err)
+	defer c.Close()
+
+	const sourceDateEpoch = int64(1234567890) // Fri Feb 13 11:31:30 PM UTC 2009
+	st := llb.Image("busybox:latest").Run(
+		llb.Args([]string{"/bin/touch", fmt.Sprintf("--date=@%d", sourceDateEpoch),
+			"/bin",
+			"/etc",
+			"/var",
+			"/var/foo",
+			"/tmp",
+			"/tmp/foo2",
+			"/tmp/foo2/bar",
+		}),
+		llb.AddMount("/var/foo", llb.Scratch(), llb.Tmpfs()),
+		llb.AddMount("/tmp/foo2/bar", llb.Scratch(), llb.Tmpfs()),
+	)
+	def, err := st.Marshal(sb.Context())
+	require.NoError(t, err)
+
+	tmpDir := t.TempDir()
+	tarFile := filepath.Join(tmpDir, "out.tar")
+	tarFileW, err := os.Create(tarFile)
+	require.NoError(t, err)
+	defer tarFileW.Close()
+
+	_, err = c.Solve(sb.Context(), def, SolveOpt{
+		Exports: []ExportEntry{
+			{
+				Type:   ExporterTar,
+				Output: fixedWriteCloser(tarFileW),
+			},
+		},
+	}, nil)
+	require.NoError(t, err)
+	tarFileW.Close()
+
+	tarFileR, err := os.Open(tarFile)
+	require.NoError(t, err)
+	defer tarFileR.Close()
+	tarR := tar.NewReader(tarFileR)
+	touched := map[string]*tar.Header{
+		"bin/": nil, // Regular dir
+		"etc/": nil, // Parent of file mounts (etc/{resolv.conf, hosts})
+		"var/": nil, // Parent of dir mount (var/foo/)
+		"tmp/": nil, // Grandparent of dir mount (tmp/foo2/bar/)
+		// No support for reproducing the timestamps of mount point directories such as var/foo/ and tmp/foo2/bar/,
```

Comment:
Could you try fixing this? https://github.com/moby/buildkit/pull/3149#discussion_r1022347004 to check the parents makes sense? Otherwise, we at least need to track it as a separate bug(hopefully in the same milestone).

Reply:
I'm not sure how it would be possible.
Maybe, we would need to modify runc to mkdir the mountpoints with the specified `SOURCE_DATE_EPOCH`.

An alternative way would be to add a new exporter opt (in a separate PR) to support `if tm > SOURCE_DATE_EPOCH { tm = SOURCE_DATE_EPOCH }`

Reply:
Not sure I understand. Isn't it just that in https://github.com/moby/buildkit/blob/09b461389b283f7268a4156df9bd2f5b9611c21a/executor/stubs.go#L32 we also need to check for the parent of that path and add it to `paths` if it doesn't exist?

Reply:
The timestamp specified in `RUN touch` is lost when the executor container exits, so it seems hard to retain custom timestamps.

`SOURCE_DATE_EPOCH` support for mount point dirs (and any arbitrary dirs/files) will be covered in
- https://github.com/containerd/containerd/pull/7710